### PR TITLE
Migrate to grpc (part1)

### PR DESCRIPTION
**Pull request by ashishsnigam**
Work on https://github.com/iotexproject/iotex-election/issues/1

1) Added new folder explorer_pb to have protobuf specific code, to avoid name collision with idl based code in folder explorer
2) Added new folders server and client to have server and client code calling for all protobuf services
3) Worked on point 1 and 2 of above github issue

### CODE REVIEW

File: explorer_pb/explorer_server.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package explorer_pb
+
+import (
+	"github.com/iotexproject/iotex-election/committee"
```
Comment (raullenchai) [inline explorer_pb/explorer_server.go]: group imports like https://github.com/iotexproject/iotex-election/blob/master/carrier/carrier.go

File: server/explorer_server.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package main
+
+import (
+	"google.golang.org/grpc"
```
Comment (raullenchai) [inline server/explorer_server.go]: group imports like https://github.com/iotexproject/iotex-election/blob/master/carrier/carrier.go

File: explorer_pb/explorer_server.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package explorer_pb
```
Comment (CoderZhi) [inline explorer_pb/explorer_server.go]: this file could be moved to "server" directory

File: server/explorer_server.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package main
```
Comment (CoderZhi) [inline server/explorer_server.go]: rename package to "server"

File: explorer_pb/explorer_server.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	"golang.org/x/net/context"
+)
+
+// server is used to implement pb.ExplorerServer.
+type server struct {
+	savedCustomers []*ExplorerServer
+	Impl
+}
+
+// Impl implements Explorer interface
```
Comment (CoderZhi) [inline explorer_pb/explorer_server.go]: Impl could be merged into "server"

File: server/explorer_server.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+func main() {
+	lis, err := net.Listen("tcp", ":8089")
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
+	// Creates a new gRPC server
+	s := grpc.NewServer()
+	pb.RegisterExplorerServer(s, &pb.server{})
```
Comment (CoderZhi) [inline server/explorer_server.go]: 1. pb.server is supposed to be a private struct in pb.
2. To construct pb.server, we need to construct "committee.Committee" and feed it to server.

File: explorer_pb/explorer_server.go (hunk 1 above)
Comment (ashishsnigam) [reply explorer_pb/explorer_server.go]: done. Ran go tool goimports on file.

Comment (ashishsnigam) [reply explorer_pb/explorer_server.go]: done. Ran go tool goimports on file.

File: explorer_pb/explorer_server.go (hunk 3 above)
Comment (ashishsnigam) [reply explorer_pb/explorer_server.go]: Done, I also initially thought of it.

File: explorer_pb/explorer_server.go (hunk 5 above)
Comment (ashishsnigam) [reply explorer_pb/explorer_server.go]: done

File: server/explorer_server.go (hunk 4 above)
Comment (ashishsnigam) [reply server/explorer_server.go]: done

File: explorer_pb/explorer.proto [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+message CandidateResponse {
+	repeated Candidate candidates = 1;
+}
+
+message BucketResponse {
+	repeated Bucket buckets = 1;
+}
+
+message Empty {
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply explorer_pb/explorer.proto]: done, Thanks.

File: explorer_pb/explorer.proto [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+	// get candidates
+	rpc getCandidates(GetCandidatesRequest) returns (CandidateResponse) {}
+
+	// get buckets by candidate
+	rpc getBucketsByCandidate(GetBucketsByCandidateRequest) returns (BucketResponse) {}
+}
+
+message ChainMeta {
+	string height = 1;
```
[the remark this answers is not in the document]
Comment (CoderZhi) [reply explorer_pb/explorer.proto]: previously, height as uint64 cannot be handled properly in js.

[the remark this answers is not in the document]
Comment (ashishsnigam) [reply explorer_pb/explorer.proto]: Not sure. I took it simply from idl file named as explorer.idl and converted it to protobuf format.
I guess if that needs to be converted into int, it also needs to be changed in idl file.

File: explorer_pb/explorer.proto [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	int64 totalCandidates = 2;
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply explorer_pb/explorer.proto]: done.

File: server/explorer_server.go (hunk 6 above)
Comment (ashishsnigam) [reply server/explorer_server.go]: Need to look more into code. For now it was just a simple work on conversion of idl file to protobuf file and add a simple method. I hope I can do it after a proper code go through. Will try to do it asap.

**Comment by ashishsnigam** [comment]
Two suggestions - 
1) We can move all protobuf files in pb directory itself, provided we maintain name space collision in different generated go files. This way there won't be any need for creating a separate directory for each protobuf service.
2) For each service defined in directory server, it must have it's own testcase in same directory which will take care of starting a gRPC server and initialise it with proper pb.server struct. (for now I don't know how to do it.)

**Comment by ashishsnigam** [comment]
Incorporated few review comments and added few suggestions. Please have a look

File: explorer_pb/explorer.proto [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+// To compile the proto, run:
+//      protoc --go_out=plugins=grpc:. *.proto
+syntax = "proto3";
+package explorer_pb;
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/empty.proto";
+
+// The Explorer service definition
+service Explorer {
```
Comment (puncsky) [inline explorer_pb/explorer.proto]: @CoderZhi how about using `BeaconChain` here?

**Pull request merged**